docs/contributor/memory_management.md

```diff
@@ -0,0 +1,182 @@
+
+At a technical level, this is represented in-memory in the following way, for an n-byte heap object:
+
+```
+ [ 32-bit counter ][ 32-bit padding ][ n-byte payload ]
+ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
+ {start address}                    {pointer used by Grain runtime}
+```
+
+Here, `{start address}` is the value returned by `Malloc.malloc` (and passed to `Malloc.free` when the object is cleaned up), and the `{pointer used by Grain runtime}` is what everything outside of the `GC` wrapper uses as the pointer to the object. In other words, we "hide" the reference count in the 8 bytes prior to the pointer.
+
+The interface provided by the `GC` module is similar (but not identical) to that which is provided by the `Malloc` module:
+
+```grain
+/**
+ * Allocates the requested number of bytes, returning a pointer with a reference count of 1.
+ *
+ * @param size: WasmI32 - The number of bytes to allocate
+ * @return WasmI32 - The pointer to the allocated region
+ */
+export let malloc = (size: WasmI32) -> WasmI32
+
+/**
+ * Frees the given pointer. Using this pointer after it has been freed will result in undefined behavior.
+ *
+ * @param userPtr: WasmI32 - The pointer to free
+ * @return Void
+ */
+export let free = (userPtr: WasmI32) -> Void
+
+/**
+ * Increments the reference count of the given pointer.
+ *
+ * @param userPtr: WasmI32 - The pointer whose reference count should be incremented
+ * @return WasmI32 - The given pointer
+ */
+export let incRef = (userPtr: WasmI32) -> WasmI32
+
+/**
+ * Decrements the reference count of the given pointer if it is greater than zero
+ *
+ * @param userPtr: WasmI32 - The pointer whose reference count should be decremented
+ * @return WasmI32 - The given pointer
+ */
+export let decRefIgnoreZeros = (userPtr: WasmI32) -> WasmI32
+
+/**
+ * Decrements the reference count of the given pointer. An error is thrown if the
+ * reference count is not greater than zero.
+ *
+ * @param userPtr: WasmI32 - The pointer whose reference count should be decremented
+ * @return WasmI32 - The given pointer
+ */
+export let decRef = (userPtr: WasmI32) -> WasmI32
+```
+
+When writing code in the runtime, `decRefIgnoreZeros` should never be used. It exists solely to support the cleanup step of compiled functions. The reference count-managing functions are safe to use with non-pointers; if a non-pointer is passed
+to them, it will be returned without any side effects. **However**, it should be noted that these functions are not safe to use with arbitrary (untagged) `WasmXX` (`WasmI32`/`WasmI64`/etc) values, which is why the use of such types is only permitted in `@disableGC` blocks (so that the compiler does not insert `incRef`/`decRef` calls).
```

Comment:
```suggestion
to them, it will be returned without any side effects. **However**, it should be noted that these functions are not safe to use with arbitrary (untagged) `WasmXX` (`WasmI32`/`WasmI64`/etc) values, which is why the use of such types is only permitted in `@disableGC` blocks—so that the compiler does not insert `incRef`/`decRef` calls.
```